### ISSUE TITLE
CPU Usage Spikes every 3-5m

### ISSUE THREAD

**Issue by bwees** [louislam/uptime-kuma#2249]
### ⚠️ Please verify that this bug has NOT been raised before.

- [X] I checked and didn't find similar issue

### 🛡️ Security Policy

- [X] I agree to have read this project [Security Policy](https://github.com/louislam/uptime-kuma/security/policy)

### Description

<img width="780" alt="Screen Shot 2022-10-19 at 9 52 33 PM" src="https://user-images.githubusercontent.com/12686250/196837634-b014b1d2-e5f1-4e70-8bf6-5526df6cc72f.png">

There are very high CPU usage spikes every 3-5 minutes. Does this align with any database cleanup tasks that run and how do I disable or lengthen the time between those operations?

### 👟 Reproduction steps

I have a mixture of TCP pings and HTTP pings. Nothing out of the ordinary. I also have a Prometheus query that runs every 1 minute to get the statistics from UptimeKuma

### 👀 Expected behavior

I would expect maybe a few spikes every hour or so or no spikes at all (after all its just pinging a web page)

### 😓 Actual Behavior

I am getting spiky CPU usage every 3 minutes that is maxing out 1 core.

### 🐻 Uptime-Kuma Version

1.18.5 but this has progressed across multiple versions

### 💻 Operating System and Arch

Raspberry Pi, arm

### 🌐 Browser

NA

### 🐋 Docker Version

Docker 20.10.13

### 🟩 NodeJS Version

NA

**Comment by louislam** [comment]
Need detailed reproduce steps

**Comment by rezzorix** [comment]
Maybe it happens when all monitors trigger and stats are pulled at exactly the same time?

**Comment by bwees** [comment]
Ok I will adjust my intervals and see if it still happens

Brandon Wees

**Comment by AuthorShin** [comment]
I'm having the same issue.

**Comment by bwees** [comment]
AuthorShin, how do you have your bind mounts set up in docker? Mine is set to a NFS share for mass storage. Not sure if having the extra latency is causing issues?

Brandon Wees

**Comment by AuthorShin** [comment]
@bwees Hey, mine is just with a simple bind mount to the local drive of the VM that is hosting the Debian 11.5 which got Docker running on which is also hosted by ESXi.
I also don't think that the shares are the problem here cause there's not much write load from Uptime Kuma no matter how many monitors you got.

**Comment by AuthorShin** [comment]
And having this spikes on CPU usage was the only reason which prevented me from installing Uptime Kuma on more environments cause there are some alerts on Prometheus/Grafana that will trigger by those spikes every few seconds.

**Comment by AuthorShin** [comment]
Is this related?
https://github.com/louislam/uptime-kuma/issues/2012

**Comment by bwees** [comment]
I have dropped my database retention to 1 day. I remember the ui freezing was due to the retention length so that seemed to fix it. Are you using the Prometheus export functionality @AuthorShin? That’s the only thing that aligns with my cpu spikes. I have changed all of my update intervals to be different multiples of 5 minutes and it still exhibits the same behavior. Not sure if there is an inefficiency in that part of the code.

**Comment by AuthorShin** [comment]
@bwees Yes I have some monitors on Prometheus as well and I use node exporter for gathering data.

**Comment by bwees** [comment]
So you are using the /metrics endpoint exposed by uptimekuma to gather data in addition to your node exporters?

**Comment by AuthorShin** [comment]
@bwees No, I use node exporters to gather data about hardware and etc and send that to Prometheus which I connect to Grafana to show those data as graphs on dashboards, on the other hand I just use Uptime Kuma to ping some IP addresses and also some TCP port along with HTTP/S endpoints of my other servers so I can see their response time / network latency and also the uptime of their network / server itself.

**Comment by bwees** [comment]
hm so it cant be the metrics endpoint. Do you have a graph of the CPU usage for the container? This is mine. Lets compare to see what patterns emerge. 
<img width="917" alt="Screen Shot 2022-11-26 at 8 52 33 PM" src="https://user-images.githubusercontent.com/12686250/204115676-40824f6f-f64e-4b45-86bf-a07c56d4b08d.png">

**Comment by bwees** [comment]
The spikes are about ever 8-10 minutes

**Comment by bwees** [comment]
Checking logs, all I get is this:

```
2022-11-27T01:20:52.205Z [RATE-LIMIT] INFO: remaining requests: 20
2022-11-27T01:21:57.221Z [RATE-LIMIT] INFO: remaining requests: 20
2022-11-27T01:23:02.204Z [RATE-LIMIT] INFO: remaining requests: 20
2022-11-27T01:24:07.210Z [RATE-LIMIT] INFO: remaining requests: 20
2022-11-27T01:25:12.206Z [RATE-LIMIT] INFO: remaining requests: 20
2022-11-27T01:26:17.209Z [RATE-LIMIT] INFO: remaining requests: 20
2022-11-27T01:27:22.204Z [RATE-LIMIT] INFO: remaining requests: 20
2022-11-27T01:28:27.210Z [RATE-LIMIT] INFO: remaining requests: 20
2022-11-27T01:29:32.204Z [RATE-LIMIT] INFO: remaining requests: 20
2022-11-27T01:30:37.209Z [RATE-LIMIT] INFO: remaining requests: 20
2022-11-27T01:31:42.204Z [RATE-LIMIT] INFO: remaining requests: 20
2022-11-27T01:32:47.210Z [RATE-LIMIT] INFO: remaining requests: 20
2022-11-27T01:33:52.205Z [RATE-LIMIT] INFO: remaining requests: 20
2022-11-27T01:34:57.222Z [RATE-LIMIT] INFO: remaining requests: 20
2022-11-27T01:36:02.204Z [RATE-LIMIT] INFO: remaining requests: 20
2022-11-27T01:37:07.209Z [RATE-LIMIT] INFO: remaining requests: 20
2022-11-27T01:38:12.205Z [RATE-LIMIT] INFO: remaining requests: 20
2022-11-27T01:39:17.209Z [RATE-LIMIT] INFO: remaining requests: 20
2022-11-27T01:40:22.205Z [RATE-LIMIT] INFO: remaining requests: 20
2022-11-27T01:41:27.210Z [RATE-LIMIT] INFO: remaining requests: 20
2022-11-27T01:42:32.205Z [RATE-LIMIT] INFO: remaining requests: 20
2022-11-27T01:43:37.209Z [RATE-LIMIT] INFO: remaining requests: 20
2022-11-27T01:44:42.206Z [RATE-LIMIT] INFO: remaining requests: 20
2022-11-27T01:45:47.210Z [RATE-LIMIT] INFO: remaining requests: 20
2022-11-27T01:46:52.204Z [RATE-LIMIT] INFO: remaining requests: 20
2022-11-27T01:47:57.210Z [RATE-LIMIT] INFO: remaining requests: 20
2022-11-27T01:49:02.205Z [RATE-LIMIT] INFO: remaining requests: 20
2022-11-27T01:50:07.209Z [RATE-LIMIT] INFO: remaining requests: 20
2022-11-27T01:51:12.204Z [RATE-LIMIT] INFO: remaining requests: 20
2022-11-27T01:52:17.209Z [RATE-LIMIT] INFO: remaining requests: 20
2022-11-27T01:53:22.204Z [RATE-LIMIT] INFO: remaining requests: 20
...
```

**Comment by AuthorShin** [comment]
Is it logs for Uptime Kuma container?

**Comment by bwees** [comment]
Yes, I am compiling a list of the watch types I use now. Will report back in a sec

**Comment by bwees** [comment]
Watch Types I Use:
- TCP Ping
- Ping
- HTTP
- DNS
- Push

**Comment by bwees** [comment]
My database is only 10 MB, not sure how often VACUUM gets run on the database. @louislam do you have an idea for this? I have gone through all of the graphs provided by the Node Exporter dashboard in grafana and I do not see any stats on any issues with disk or something like that. I think I may `top` inside the container and see if any processes come up.

**Comment by AuthorShin** [comment]
My database is even less :D

**Comment by bwees** [comment]
lol, yeah all I can see process wise is a spike in the node process but that's it :/

**Comment by bwees** [comment]
interesting.... I still think it's related to the vacuum process in sqlite being triggered because it looks like you have lower CPU spikes which could be related to a smaller database size

**Comment by AuthorShin** [comment]
Maybe. do you got spikes on RAM too?

**Comment by AuthorShin** [comment]
My container logs are usually containing of these things:

`
2022-11-26T19:54:06.365Z [MONITOR] INFO: Get Monitor Beats: 4 User ID: 1

2022-11-26T19:54:15.652Z [MONITOR] INFO: Get Monitor Beats: 3 User ID: 1

2022-11-26T19:54:19.511Z [MONITOR] INFO: Get Monitor Beats: 3 User ID: 1

2022-11-26T19:54:28.563Z [MONITOR] INFO: Get Monitor Beats: 1 User ID: 1

2022-11-26T19:54:46.175Z [MONITOR] INFO: Get Monitor Beats: 2 User ID: 1

2022-11-26T19:58:27.094Z [MONITOR] INFO: Get Monitor Beats: 3 User ID: 1

2022-11-26T19:58:29.685Z [MONITOR] INFO: Get Monitor Beats: 5 User ID: 1

2022-11-26T19:58:31.186Z [MONITOR] INFO: Get Monitor Beats: 4 User ID: 1


`

**Comment by bwees** [comment]
Interesting, i think my rate limits might be from my Prometheus query every 60s. Looking into the docs for SQLite, if auto_vacuum is set to full, then it will run a vacuum command for every transaction. Lemme see if I can find that setting int he code rq.

I also see a spike in memory usage similar to your picture.

**Comment by bwees** [comment]
Found it
https://github.com/louislam/uptime-kuma/blob/cc6d17d2e0e6b28b696ee96030251701a2bdcabe/server/database.js#L153

**Comment by AuthorShin** [comment]
Hmmm interesting, so for each PING or beat (write on database) on any monitor there was a vacuum trying to clean the database as well?

**Comment by rezzorix** [comment]
Guys, you can discuss this back and forth…

Your error comes from using NFS shares etc. 

This is not supported; see main uptime-kuma github page as reference, there is even a ⚠️ in front as a warning as it will cause problems.

**Comment by AuthorShin** [comment]
@rezzorix I'm not using NFS shares etc.

**Comment by rezzorix** [comment]
Ah damn, I just read on mobile and from that I somehow got the impression you guys all used NFS shares.

So lets get back to the basics:
Just the simple container without any special configs or other apps trying to access db and some monitors.
Do you still have this spikes?

I run several instances of uptime-kuma. 1 on a raspberry pi zero2 with an SD card...
...aside from of course the loading of stats taking sweet time; with around 50 monitors (mainly HTTPS keyword, TCP ping, 60/2/30) I cannot see much load or even spikes on the CPU.

**Comment by AuthorShin** [comment]
@rezzorix Yeah even with very basic setup and 4 Ping monitors

**Comment by bwees** [comment]
@AuthorShin maybe set it with 4 http monitors and see if its the ping monitor that is causing it?

**Comment by louislam** [comment]
> Found it
> 
> https://github.com/louislam/uptime-kuma/blob/cc6d17d2e0e6b28b696ee96030251701a2bdcabe/server/database.js#L153

You can try to test it by forking this project and change it to `NONE` and run the container using this special image:

```
docker run --rm -it -p 3000:3000 -p 3001:3001 --pull always -e 'UPTIME_KUMA_GH_REPO=<REPO>' louislam/uptime-kuma:pr-test
```

Where <REPO> = `bwees:master`


Reference: https://github.com/louislam/uptime-kuma/wiki/Test-Pull-Requests

**Comment by chakflying** [comment]
For container healthcheck, a seperate `node` process is launched every minute to run the healthcheck script. I have found that when nothing is running this can cause CPU spikes.

![Screenshot](https://user-images.githubusercontent.com/3271800/204227653-97dc3115-6ecb-4b3a-83e2-3a9308958006.png)

One can test if this is the issue by creating the container with additional argument `--no-healthcheck`:

```shell
docker run -d --restart=always -p 3001:3001 -v uptime-kuma:/app/data --no-healthcheck --name uptime-kuma louislam/uptime-kuma:1
```

**Comment by AuthorShin** [comment]
Thanks @chakflying 
Your solution seems to be working for me.
But one question is there a way to add `--no-healthcheck` as an environment variables?

<img src="https://i.ibb.co/sg4wVKf/Screenshot-2022-11-28-163902.png" alt="Screenshot-2022-11-28-163902">

**Comment by chakflying** [comment]
I don't think it's possible since healthcheck is controlled by docker (daemon?).

Switching to a more simple healthcheck, maybe a shell script, should be a better solution in the long run.

**Comment by louislam** [comment]
Great finding, currently the healthcheck script is written in JS which is required cool launching of Node.js.

https://github.com/louislam/uptime-kuma/blob/28d13e198c103f30be0d697e58be3779036b9c67/docker/dockerfile#L54

https://github.com/louislam/uptime-kuma/blob/master/extra/healthcheck.js

**Comment by bwees** [comment]
> Thanks @chakflying Your solution seems to be working for me. But one question is there a way to add `--no-healthcheck` as an environment variables?
> 
> ![Screenshot-2022-11-28-163902](https://camo.githubusercontent.com/249a597bc78361223c3bd1017f79aa237cbd3844c794f39bd4f221e357d99ae0/68747470733a2f2f692e6962622e636f2f73673477564b662f53637265656e73686f742d323032322d31312d32382d3136333930322e706e67)

are you running from docker compose? You can just add 
```
healthcheck: 
  disable: true
``` 
to your service like this:

```
  uptimekuma:
    image: louislam/uptime-kuma
    healthcheck:
      disable: true
```

**Comment by AuthorShin** [comment]
@bwees I'm using CasaOS as my docker manager and currently it doesn't support that that's why I was asking for environment variables.

**Comment by louislam** [comment]
I have rewritten the healthcheck in go-lang which should be very close to native performance.

**Comment by thoompje** [comment]
> > Thanks @chakflying Your solution seems to be working for me. But one question is there a way to add `--no-healthcheck` as an environment variables?
> > ![Screenshot-2022-11-28-163902](https://camo.githubusercontent.com/249a597bc78361223c3bd1017f79aa237cbd3844c794f39bd4f221e357d99ae0/68747470733a2f2f692e6962622e636f2f73673477564b662f53637265656e73686f742d323032322d31312d32382d3136333930322e706e67)
> 
> are you running from docker compose? You can just add
> 
> ```
> healthcheck: 
>   disable: true
> ```
> 
> to your service like this:
> 
> ```
>   uptimekuma:
>     image: louislam/uptime-kuma
>     healthcheck:
>       disable: true
> ```

Where to edit this? 

plenty of messages and kuma is so slow here

2023-10-31T08:59:51+01:00 [RATE-LIMIT] INFO: remaining requests: 20
2023-10-31T09:00:03+01:00 [RATE-LIMIT] INFO: remaining requests: 20
2023-10-31T09:00:25+01:00 [RATE-LIMIT] INFO: remaining requests: 20
2023-10-31T09:00:44+01:00 [RATE-LIMIT] INFO: remaining requests: 20

**Comment by chakflying** [comment]
The original issue was fixed a long time ago, any slow down you experience would not be related to this issue.
